model/network_user_model.go

```diff
@@ -57,7 +84,780 @@ func GetNetworkUser(
+			returnErr = err
+			return
+		}
+
+		/**
+		 * No record exists with this auth type, create a new one
+		 */
+
+		_, dbErr := tx.Exec(
+			ctx,
+			`
+				INSERT INTO network_user_auth_password
+				(user_id, user_auth, auth_type, password_salt, password_hash, verified)
+				VALUES ($1, $2, $3, $4, $5, $6)
+			`,
+			args.UserId,
+			userAuth,
+			userAuthType,
+			args.PasswordSalt,
+			args.PasswordHash,
+			args.Verified,
+		)
+		server.Raise(dbErr)
+	})
+
+	return
+}
+
+func getUserAuths(
+	userId server.Id,
+	ctx context.Context,
+) ([]NetworkUserUserAuth, error) {
+
+	var userAuths []NetworkUserUserAuth
+
+	server.Tx(ctx, func(tx server.PgTx) {
+
+		result, err := tx.Query(
+			ctx,
+			`
+			SELECT
+				user_auth,
+				auth_type,
+				password_hash,
+				password_salt
+			FROM network_user_auth_password
+			WHERE user_id = $1
+		`,
+			userId,
+		)
+		if err != nil {
+			server.Raise(err)
+		}
+
+		server.WithPgResult(result, err, func() {
+			for result.Next() {
+				userAuth := NetworkUserUserAuth{}
+				server.Raise(result.Scan(
+					&userAuth.UserAuth,
+					&userAuth.AuthType,
+					&userAuth.PasswordHash,
+					&userAuth.PasswordSalt,
+				))
+				userAuths = append(userAuths, userAuth)
+			}
+		})
+
+	})
+
+	return userAuths, nil
+
+}
+
+/**
+ * Allow different SSO auth methods
+ */
+
+type AddSsoAuthArgs struct {
+	ParsedAuthJwt AuthJwt     `json:"auth_jwt"`
+	AuthJwt       string      `json:"auth_jwt_str"`
+	AuthJwtType   SsoAuthType `json:"auth_jwt_type"`
+	UserId        server.Id   `json:"user_id"`
+}
+
+func addSsoAuth(
+	args *AddSsoAuthArgs,
+	ctx context.Context,
+) (returnErr error) {
+
+	parsedAuthJwt := args.ParsedAuthJwt
+
+	normalJwtUserAuth, _ := NormalUserAuth(parsedAuthJwt.UserAuth)
+
+	server.Tx(ctx, func(tx server.PgTx) {
+
+		/**
+		 * first, check if the user_auth is already associated with a different user
+		 */
+		query, err := tx.Query(
+			ctx,
+			`
+			SELECT COUNT(*)
+		    FROM (
+		        SELECT user_id
+		        FROM network_user_auth_sso
+		        WHERE user_auth = $1 AND user_id <> $2
+
+		        UNION
+
+		        SELECT user_id
+		        FROM network_user_auth_password
+		        WHERE user_auth = $1 AND user_id <> $2
+		    ) different_users
+			`,
+			parsedAuthJwt.UserAuth,
+			args.UserId,
+		)
+
+		if err != nil {
+			glog.Errorf("Error querying for user auth conflicts: %s", err.Error())
+			returnErr = err
+			return
+		}
+
+		conflictCount := 0
+
+		server.WithPgResult(query, err, func() {
+			if query.Next() {
+				server.Raise(query.Scan(&conflictCount))
+			}
+		})
+
+		if conflictCount > 0 {
+			returnErr = fmt.Errorf("user_auth %s already exists for a different user", parsedAuthJwt.UserAuth)
+			return
+		}
+
+		/**
+		 * No conflicts, insert the new SSO auth
+		 */
+
+		result, err := tx.Exec(
+			ctx,
+			`
+			INSERT INTO network_user_auth_sso
+			(user_id, auth_type, user_auth, auth_jwt)
+			VALUES ($1, $2, $3, $4)
+			ON CONFLICT (user_id, auth_type) DO NOTHING;
```

Comment:
`ON CONFLICT (user_auth, auth_type)` would happen with a UNIQUE index. In that case, DO NOTHING. You can check `result.RowsAffected()` to see if the row was inserted.